### PR TITLE
helm chart for hypershift

### PR DESCRIPTION
**Pull request by geoberle**
### What this PR does

use native `hypershift install` functionality within helm chart
the external DNS part is still part of this helm chart. will be removed during P3.

Jira: https://issues.redhat.com/browse/ARO-11084
Link to demo recording:

### CODE REVIEW

File: hypershiftoperator/deploy/helm/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v2
+description: A Helm chart to install the Hypershift Operator and deps for ARO
+name: aor-hcp-hypershift-operator
```
Comment (tony-schndr) [inline hypershiftoperator/deploy/helm/Chart.yaml:3]: typo

File: hypershiftoperator/deploy/helm/charts/external-dns/templates/deployment-external-dns.yaml [hunk 2]
```diff
@@ -1,70 +1,69 @@
...
+        - --registry=txt
+        - --txt-suffix=-external-dns
+        - --txt-owner-id={{ .Values.txtOwnerId }}
+        - --label-filter=hypershift.openshift.io/route-visibility!=private
+        - --interval=1m
+        - --txt-cache-interval=1h
+        - --azure-config-file=/etc/provider/credentials
+        command:
+        - /ko-app/external-dns
+        image: "{{ .Values.image }}:{{ .Values.imagetag }}"
```
Comment (tony-schndr) [inline hypershiftoperator/deploy/helm/charts/external-dns/templates/deployment-external-dns.yaml:37]: nit: camelCase for `.Values.imageTag`

File: hypershiftoperator/deploy/helm/templates/installer.job.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
+spec:
+  # set deadline to 30min
+  activeDeadlineSeconds: 1800
+  backoffLimit: 1
+  template:
+    spec:
+      containers:
+        - name: install
+          image: "{{ .Values.image }}:{{ .Values.imagetag }}"
```
Comment (tony-schndr) [inline hypershiftoperator/deploy/helm/templates/installer.job.yaml:18]: nit: camelCase `imageTag`

File: hypershiftoperator/deploy/helm/templates/installer.job.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+          command:
+            - /bin/sh
+            - -c
+            - |
+              hypershift install \
+                --enable-conversion-webhook=false \
+                --managed-service ARO-HCP \
+                --aro-hcp-key-vault-users-client-id {{ .Values.azureKeyVaultClientId }} \
+                --registry-overrides "{{ .Values.registryOverrides }}" \
+                --hypershift-image {{ .Values.image }}:{{ .Values.imagetag }} \
```
Comment (tony-schndr) [inline hypershiftoperator/deploy/helm/templates/installer.job.yaml:28]: same nit as above

**Review by tony-schndr — COMMENTED** [review]
overall LGTM, just a few nits/questions.

File: config/public-cloud-personal-dev.json [hunk 5]
```diff
@@ -14,14 +14,17 @@
...
   "externalDNSImageTag": "v0.14.2",
+  "externalDNSManagedIdentityName": "external-dns",
+  "externalDNSServiceAccountName": "external-dns",
   "firstPartyAppClientId": "57e54810-3138-4f38-bd3b-29cb33f4c358",
   "frontendCosmosDBDeploy": true,
   "frontendCosmosDBDisableLocalAuth": true,
   "frontendCosmosDBName": "aro-hcp-rp-76fc6",
   "globalRG": "global",
   "grafanaAdminGroupPrincipalId": "6b6d3adf-8476-4727-9812-20ffdef2b85c",
   "grafanaName": "aro-hcp-grafana-76fc6",
```
[the remark this answers is not in the document]
Comment (geoberle) [reply config/public-cloud-personal-dev.json:25]: to force people to review the effect a change on config.yaml has. really no other reason as these files are not used for anything else.

File: hypershiftoperator/deploy/helm/Chart.yaml (hunk 1 above)
Comment (geoberle) [reply hypershiftoperator/deploy/helm/Chart.yaml]: ah yes, the lesser known AOR service :) 
fixed

File: hypershiftoperator/Makefile [hunk 6]
```diff
@@ -3,69 +3,24 @@ DEPLOY_ENV ?= personal-dev
...
-.PHONY: create-edo-azure-creds create-edo-sa-patch create-domain-file create-txt-owner-id-file deploy unpack-ho-manifests
+deploy:
+	@EXTERNAL_DNS_OPERATOR_MI_CLIENT_ID=$(shell az identity show -g ${RESOURCEGROUP} -n ${EXTERNAL_DNS_MI_NAME} --query clientId -o tsv) && \
+	AZURE_TENANT_ID=$(shell az account show --query tenantId --output tsv) && \
+	AZURE_SUBSCRIPTION_ID=$(shell az account show --query id --output tsv) && \
+	CSI_SECRET_STORE_CLIENT_ID=$(shell az aks show -n ${AKS_NAME} -g ${RESOURCEGROUP} --query 'addonProfiles.azureKeyvaultSecretsProvider.identity.clientId' -o tsv) && \
+	helm upgrade --install hypershift deploy/helm \
+		--create-namespace --namespace ${HYPERSHIFT_NAMESPACE} \
+		--set image=${HO_IMAGE_BASE} \
+		--set imagetag=${HO_IMAGE_TAG} \
```
Comment (tony-schndr) [inline hypershiftoperator/Makefile:14]: is `imageTag` in values.yaml

File: hypershiftoperator/Makefile [hunk 7]
```diff
@@ -3,69 +3,24 @@ DEPLOY_ENV ?= personal-dev
...
+		--set registryOverrides="quay.io/openshift-release-dev/ocp-v4.0-art-dev=${ARO_HCP_OCP_ACR}.azurecr.io/openshift/release\,quay.io/openshift-release-dev/ocp-release=${ARO_HCP_OCP_ACR}.azurecr.io/openshift/release-images\,registry.redhat.io/redhat=${ARO_HCP_OCP_ACR}.azurecr.io/redhat" \
+		--set azureKeyVaultClientId=$${CSI_SECRET_STORE_CLIENT_ID} \
+		--set external-dns.image=${ED_IMAGE_BASE} \
+		--set external-dns.imagetag=${ED_IMAGE_TAG} \
```
Comment (tony-schndr) [inline hypershiftoperator/Makefile:18]: is `imageTag` in values.yaml

**Comment by tony-schndr** [comment]
When deploying this the servicemonitor is created under the wrong group, should be `azmonitoring.coreos.com`
```
kubectl get servicemonitor.monitoring.coreos.com -A
NAMESPACE                                                       NAME                                  AGE
hypershift                                                      operator                              18h
```

File: hypershiftoperator/Makefile (hunk 6 above)
Comment (geoberle) [reply hypershiftoperator/Makefile]: good spot. ty

**Comment by geoberle** [comment]
> When deploying this the servicemonitor is created under the wrong group, should be `azmonitoring.coreos.com`
> 
> ```
> kubectl get servicemonitor.monitoring.coreos.com -A
> NAMESPACE                                                       NAME                                  AGE
> hypershift                                                      operator                              18h
> ```

we can't avoid that hypershift installs the `monitoring.coreos.com` CRs for the time being, but i've added an `azmonitoring.coreos.com` flavor of the `operator` `ServiceMonitor` so we can scrape with managed prometheus.

**Review by tony-schndr — APPROVED** [review]

**Pull request merged**